src/engine/providers/openAi.ts

```diff
@@ -23,38 +24,22 @@ const OpenAi: Provider = {
   },
 
   getChatCompletion: async (config: ProviderConfig, messages: Message[]) => {
-    const openai = new OpenAI({
+    const api = new OpenAI({
       apiKey: config.apiKey,
     });
 
-    const systemMessage: Message = {
-      role: 'system',
-      content: config.systemPrompt,
-    };
+    return await getChatCompletion(api, config, messages);
+  },
 
-    const startTime = performance.now();
-    const response = await openai.chat.completions.create({
-      messages: [systemMessage, ...messages],
-      model: config.model,
-      ...responseStyles[config.responseStyle],
+  getChatCompletionStream: async function* (
+    config: ProviderConfig,
+    messages: Message[],
+  ): AsyncGenerator<ModelResponseStream> {
+    const api = new OpenAI({
+      apiKey: config.apiKey,
     });
 
-    const responseTime = performance.now() - startTime;
-
-    return {
-      message: {
-        role: 'assistant',
-        content: response.choices[0]?.message.content ?? '',
-      },
-      usage: {
-        inputTokens: response.usage?.prompt_tokens ?? 0,
-        outputTokens: response.usage?.completion_tokens ?? 0,
-        requests: 1,
-      },
-      responseTime,
-      responseModel: response.model,
-      data: response,
-    };
+    yield* getBetaChatCompletionStream(api, config, messages);
```

Comment:
how about putting it behind a flag?

Reply:
I plan to enable streaming by default, as it can be accessed using "non-beta" functions. The issue here, is that streamed responses from OpenAI-type API (OpenAI, perplexity) lack the `usage` field which informated about used tokens and was used to calculate costs. Non-streamed responses have it, streamed responses don't 🤦‍♂️

I plan to add `--no-stream` CLI option to disable streaming. In order to compenase for lack for usage information comming from API we might resort to calculate usage tokens locally, however that won't be 100% accurate, especially for perplexity models as we might not be able to get the exact tokenizer for them.